File: nasdaq-cse/oms/manager.py
"""
Order Management System (OMS) for handling order lifecycle
"""
import uuid
from datetime import datetime
from typing import List, Dict, Optional
from sqlalchemy.orm import Session
from core.models import Order, Trade, Position, Contract, OrderSide, OrderType, OrderStatus
from storage.database import db_manager
import asyncio
import random


class MatchingEngine:
    """
    Simple matching engine for order execution
    """
    
    def __init__(self):
        self.order_book = {
            'bids': [],  # Buy orders sorted by price (highest first)
            'asks': []   # Sell orders sorted by price (lowest first)
        }
        self.last_trade_price = 2050.0
    
    async def process_order(self, order: Order, db: Session) -> List[Trade]:
        """
        Process an order through the matching engine
        """
        trades = []
        
        if order.order_type == OrderType.MARKET:
            trades = await self._execute_market_order(order, db)
        elif order.order_type == OrderType.LIMIT:
            trades = await self._execute_limit_order(order, db)
        
        return trades
    
    async def _execute_market_order(self, order: Order, db: Session) -> List[Trade]:
        """
        Execute market order at current market price
        """
        # Simulate market execution with small slippage
        execution_price = self.last_trade_price * (1 + random.uniform(-0.001, 0.001))
        
        trade = Trade(
            trade_id=str(uuid.uuid4()),
            buy_order_id=order.order_id if order.side == OrderSide.BUY else None,
            sell_order_id=order.order_id if order.side == OrderSide.SELL else None,
            contract_id=order.contract_id,
            quantity=order.quantity,
            price=execution_price,
            trade_time=datetime.utcnow()
        )
        
        db.add(trade)
        
        # Update order status
        order.status = OrderStatus.FILLED
        order.filled_quantity = order.quantity
        order.avg_fill_price = execution_price
        order.updated_at = datetime.utcnow()
        
        self.last_trade_price = execution_price
        
        return [trade]
    
    async def _execute_limit_order(self, order: Order, db: Session) -> List[Trade]:
        """
        Execute limit order if price conditions are met
        """
        # Simplified limit order execution
        # In real system, this would check order book for matching orders
        
        current_market_price = self.last_trade_price
        can_execute = False
        
        if order.side == OrderSide.BUY and order.price >= current_market_price:
            can_execute = True
        elif order.side == OrderSide.SELL and order.price <= current_market_price:
            can_execute = True
        
        if can_execute:
            return await self._execute_market_order(order, db)
        else:
            # Order remains pending
            order.status = OrderStatus.PENDING
            return []
    
    def get_market_depth(self) -> Dict:
        """
        Get current market depth (order book)
        """
        return {
            'bids': self.order_book['bids'][:10],  # Top 10 bids
            'asks': self.order_book['asks'][:10],  # Top 10 asks
            'last_price': self.last_trade_price
        }


class OrderManager:
    """
    Manages order lifecycle and position tracking
    """
    
    def __init__(self):
        self.matching_engine = MatchingEngine()
    
    async def submit_order(self, user_id: int, order_request: Dict) -> Dict:
        """
        Submit a new order
        """
        db = next(db_manager.get_db())
        
        try:
            # Get contract
            contract = db.query(Contract).filter(
                Contract.symbol == order_request['contract_symbol']
            ).first()
            
            if not contract:
                return {'success': False, 'error': 'Contract not found'}
            
            # Create order
            order = Order(
                order_id=str(uuid.uuid4()),
                user_id=user_id,
                contract_id=contract.id,
                side=OrderSide(order_request['side']),
                order_type=OrderType(order_request['order_type']),
                quantity=float(order_request['quantity']),
                price=float(order_request.get('price', 0)) if order_request.get('price') else None,
                stop_price=float(order_request.get('stop_price', 0)) if order_request.get('stop_price') else None,
                status=OrderStatus.PENDING
            )
            
            db.add(order)
            db.commit()
            
            # Process order through matching engine
            trades = await self.matching_engine.process_order(order, db)
            
            # Update positions if order was executed
            if trades:
                await self._update_positions(user_id, trades, db)
            
            db.commit()
            
            return {
                'success': True,
                'order_id': order.order_id,
                'status': order.status.value,
                'trades': [{'trade_id': trade.trade_id, 'price': trade.price, 'quantity': trade.quantity} for trade in trades]
            }
            
        except Exception as e:
            db.rollback()
            return {'success': False, 'error': str(e)}
        finally:
            db.close()
    
    async def cancel_order(self, order_id: str, user_id: int) -> Dict:
        """
        Cancel an existing order
        """
        db = next(db_manager.get_db())
        
        try:
            order = db.query(Order).filter(
                Order.order_id == order_id,
                Order.user_id == user_id,
                Order.status == OrderStatus.PENDING
            ).first()
            
            if not order:
                return {'success': False, 'error': 'Order not found or cannot be cancelled'}
            
            order.status = OrderStatus.CANCELLED
            order.updated_at = datetime.utcnow()
            
            db.commit()
            
            return {'success': True, 'message': 'Order cancelled successfully'}
            
        except Exception as e:
            db.rollback()
            return {'success': False, 'error': str(e)}
        finally:
            db.close()
    
    async def get_user_orders(self, user_id: int, limit: int = 100) -> List[Dict]:
        """
        Get user's orders
        """
        db = next(db_manager.get_db())
        
        try:
            orders = db.query(Order).filter(Order.user_id == user_id).order_by(Order.created_at.desc()).limit(limit).all()
            
            return [
                {
                    'order_id': order.order_id,
                    'contract_id': order.contract_id,
                    'side': order.side.value,
                    'order_type': order.order_type.value,
                    'quantity': order.quantity,
                    'price': order.price,
                    'status': order.status.value,
                    'filled_quantity': order.filled_quantity,
                    'avg_fill_price': order.avg_fill_price,
                    'created_at': order.created_at.isoformat(),
                    'updated_at': order.updated_at.isoformat()
                }
                for order in orders
            ]
            
        finally:
            db.close()
    
    async def get_user_trades(self, user_id: int, limit: int = 100) -> List[Dict]:
        """
        Get user's trade history
        """
        db = next(db_manager.get_db())
        
        try:
            # Get trades through orders
            trades = db.query(Trade).join(Order, 
                (Trade.buy_order_id == Order.order_id) | (Trade.sell_order_id == Order.order_id)
            ).filter(Order.user_id == user_id).order_by(Trade.trade_time.desc()).limit(limit).all()
            
            return [
                {
                    'trade_id': trade.trade_id,
                    'contract_id': trade.contract_id,
                    'quantity': trade.quantity,
                    'price': trade.price,
                    'trade_time': trade.trade_time.isoformat()
                }
                for trade in trades
            ]
            
        finally:
            db.close()
    
    async def get_user_positions(self, user_id: int) -> List[Dict]:
        """
        Get user's current positions
        """
        db = next(db_manager.get_db())
        
        try:
            positions = db.query(Position).filter(Position.user_id == user_id).all()
            
            return [
                {
                    'position_id': pos.id,
                    'contract_id': pos.contract_id,
                    'quantity': pos.quantity,
                    'avg_entry_price': pos.avg_entry_price,
                    'unrealized_pnl': pos.unrealized_pnl,
                    'realized_pnl': pos.realized_pnl,
                    'margin_requirement': pos.margin_requirement,
                    'last_updated': pos.last_updated.isoformat()
                }
                for pos in positions
            ]
            
        finally:
            db.close()
    
    async def _update_positions(self, user_id: int, trades: List[Trade], db: Session):
        """
        Update user positions based on executed trades
        """
        for trade in trades:
            # Find existing position
            position = db.query(Position).filter(
                Position.user_id == user_id,
                Position.contract_id == trade.contract_id
            ).first()
            
            # Determine if this is a buy or sell
            order = db.query(Order).filter(
                (Order.order_id == trade.buy_order_id) | (Order.order_id == trade.sell_order_id)
            ).first()
            
            if not order:
                continue
            
            trade_quantity = trade.quantity if order.side == OrderSide.BUY else -trade.quantity
            
            if position:
                # Update existing position
                old_quantity = position.quantity
                old_value = old_quantity * position.avg_entry_price
                new_value = trade_quantity * trade.price
                
                position.quantity += trade_quantity
                
                if position.quantity != 0:
                    position.avg_entry_price = (old_value + new_value) / position.quantity
                else:
                    # Position closed
                    position.realized_pnl += old_value + new_value
                    position.avg_entry_price = 0
                
                position.last_updated = datetime.utcnow()
                
            else:
                # Create new position
                contract = db.query(Contract).filter(Contract.id == trade.contract_id).first()
                margin_req = contract.initial_margin if contract else 1000.0
                
                position = Position(
                    user_id=user_id,
                    contract_id=trade.contract_id,
                    quantity=trade_quantity,
                    avg_entry_price=trade.price,
                    margin_requirement=margin_req,
                    last_updated=datetime.utcnow()
                )
                db.add(position)
    
    async def update_position_pnl(self, current_prices: Dict[int, float]):
        """
        Update unrealized P&L for all positions based on current market prices
        """
        db = next(db_manager.get_db())
        
        try:
            positions = db.query(Position).all()
            
            for position in positions:
                if position.contract_id in current_prices:
                    current_price = current_prices[position.contract_id]
                    if position.quantity != 0:
                        position.unrealized_pnl = (current_price - position.avg_entry_price) * position.quantity
                        position.last_updated = datetime.utcnow()
            
            db.commit()
            
        finally:
            db.close()


# Global instance
order_manager = OrderManager()

File: nasdaq-cse-go/internal/oms/manager.go
// Package oms provides Order Management System functionality
package oms

import (
	"fmt"
	"math/rand"
	"time"

	"github.com/google/uuid"
	"github.com/smaruf/python-ai-course/nasdaq-cse-go/internal/core"
	"gorm.io/gorm"
)

// MatchingEngine handles order execution and matching
type MatchingEngine struct {
	orderBook     map[string][]core.Order // Key: contract symbol
	lastTradePrice map[string]float64      // Key: contract symbol
}

// NewMatchingEngine creates a new matching engine
func NewMatchingEngine() *MatchingEngine {
	return &MatchingEngine{
		orderBook:      make(map[string][]core.Order),
		lastTradePrice: make(map[string]float64),
	}
}

// ProcessOrder processes an order through the matching engine
func (me *MatchingEngine) ProcessOrder(order *core.Order, contract *core.Contract, db *gorm.DB) ([]core.Trade, error) {
	var trades []core.Trade

	switch order.OrderType {
	case core.OrderTypeMarket:
		trade, err := me.executeMarketOrder(order, contract, db)
		if err != nil {
			return nil, err
		}
		trades = append(trades, *trade)
	case core.OrderTypeLimit:
		executedTrades, err := me.executeLimitOrder(order, contract, db)
		if err != nil {
			return nil, err
		}
		trades = append(trades, executedTrades...)
	default:
		return nil, fmt.Errorf("unsupported order type: %s", order.OrderType)
	}

	return trades, nil
}

// executeMarketOrder executes a market order at current market price
func (me *MatchingEngine) executeMarketOrder(order *core.Order, contract *core.Contract, db *gorm.DB) (*core.Trade, error) {
	// Get last trade price or use default
	lastPrice, exists := me.lastTradePrice[contract.Symbol]
	if !exists {
		lastPrice = 2050.0 // Default gold price
	}

	// Simulate market execution with small slippage
	slippage := (rand.Float64() - 0.5) * 0.002 // ±0.1% slippage
	executionPrice := lastPrice * (1 + slippage)

	// Create trade
	trade := core.Trade{
		TradeID:    uuid.New().String(),
		ContractID: order.ContractID,
		Quantity:   order.Quantity,
		Price:      executionPrice,
		TradeTime:  time.Now(),
	}

	// Set order IDs based on side
	if order.Side == core.OrderSideBuy {
		trade.BuyOrderID = &order.OrderID
	} else {
		trade.SellOrderID = &order.OrderID
	}

	// Update order status
	order.Status = core.OrderStatusFilled
	order.FilledQuantity = order.Quantity
	avgPrice := executionPrice
	order.AvgFillPrice = &avgPrice

	// Save trade to database
	if err := db.Create(&trade).Error; err != nil {
		return nil, fmt.Errorf("failed to create trade: %w", err)
	}

	// Update last trade price
	me.lastTradePrice[contract.Symbol] = executionPrice

	return &trade, nil
}

// executeLimitOrder executes a limit order if price conditions are met
func (me *MatchingEngine) executeLimitOrder(order *core.Order, contract *core.Contract, db *gorm.DB) ([]core.Trade, error) {
	if order.Price == nil {
		return nil, fmt.Errorf("limit order must have a price")
	}

	// Get current market price
	lastPrice, exists := me.lastTradePrice[contract.Symbol]
	if !exists {
		lastPrice = 2050.0
	}

	canExecute := false
	if order.Side == core.OrderSideBuy && *order.Price >= lastPrice {
		canExecute = true
	} else if order.Side == core.OrderSideSell && *order.Price <= lastPrice {
		canExecute = true
	}

	if canExecute {
		// Execute at limit price
		trade := core.Trade{
			TradeID:    uuid.New().String(),
			ContractID: order.ContractID,
			Quantity:   order.Quantity,
			Price:      *order.Price,
			TradeTime:  time.Now(),
		}

		if order.Side == core.OrderSideBuy {
			trade.BuyOrderID = &order.OrderID
		} else {
			trade.SellOrderID = &order.OrderID
		}

		// Update order status
		order.Status = core.OrderStatusFilled
		order.FilledQuantity = order.Quantity
		order.AvgFillPrice = order.Price

		// Save trade
		if err := db.Create(&trade).Error; err != nil {
			return nil, fmt.Errorf("failed to create trade: %w", err)
		}

		me.lastTradePrice[contract.Symbol] = *order.Price
		return []core.Trade{trade}, nil
	}

	// Order remains pending
	order.Status = core.OrderStatusPending
	return []core.Trade{}, nil
}

// GetMarketDepth returns current market depth
func (me *MatchingEngine) GetMarketDepth(symbol string) map[string]interface{} {
	orders := me.orderBook[symbol]
	
	var bids, asks []core.Order
	for _, order := range orders {
		if order.Status == core.OrderStatusPending {
			if order.Side == core.OrderSideBuy {
				bids = append(bids, order)
			} else {
				asks = append(asks, order)
			}
		}
	}

	// Limit to top 10
	if len(bids) > 10 {
		bids = bids[:10]
	}
	if len(asks) > 10 {
		asks = asks[:10]
	}

	return map[string]interface{}{
		"bids":       bids,
		"asks":       asks,
		"last_price": me.lastTradePrice[symbol],
	}
}

// OrderManager manages order lifecycle and position tracking
type OrderManager struct {
	matchingEngine *MatchingEngine
	db             *gorm.DB
}

// NewOrderManager creates a new order manager
func NewOrderManager(db *gorm.DB) *OrderManager {
	return &OrderManager{
		matchingEngine: NewMatchingEngine(),
		db:             db,
	}
}

// SubmitOrderResult represents the result of submitting an order
type SubmitOrderResult struct {
	Success bool                     `json:"success"`
	OrderID string                   `json:"order_id,omitempty"`
	Status  string                   `json:"status,omitempty"`
	Trades  []map[string]interface{} `json:"trades,omitempty"`
	Error   string                   `json:"error,omitempty"`
}

// SubmitOrder submits a new order
func (om *OrderManager) SubmitOrder(userID uint, orderRequest core.OrderCreateRequest) SubmitOrderResult {
	// Get contract
	var contract core.Contract
	if err := om.db.Where("symbol = ?", orderRequest.ContractSymbol).First(&contract).Error; err != nil {
		return SubmitOrderResult{
			Success: false,
			Error:   "Contract not found",
		}
	}

	// Create order
	order := core.Order{
		OrderID:    uuid.New().String(),
		UserID:     userID,
		ContractID: contract.ID,
		Side:       orderRequest.Side,
		OrderType:  orderRequest.OrderType,
		Quantity:   orderRequest.Quantity,
		Price:      orderRequest.Price,
		StopPrice:  orderRequest.StopPrice,
		Status:     core.OrderStatusPending,
	}

	// Save order to database
	if err := om.db.Create(&order).Error; err != nil {
		return SubmitOrderResult{
			Success: false,
			Error:   fmt.Sprintf("Failed to create order: %v", err),
		}
	}

	// Process order through matching engine
	trades, err := om.matchingEngine.ProcessOrder(&order, &contract, om.db)
	if err != nil {
		return SubmitOrderResult{
			Success: false,
			Error:   fmt.Sprintf("Failed to process order: %v", err),
		}
	}

	// Update order in database
	if err := om.db.Save(&order).Error; err != nil {
		return SubmitOrderResult{
			Success: false,
			Error:   fmt.Sprintf("Failed to update order: %v", err),
		}
	}

	// Update positions if order was executed
	if len(trades) > 0 {
		if err := om.updatePositions(userID, trades); err != nil {
			return SubmitOrderResult{
				Success: false,
				Error:   fmt.Sprintf("Failed to update positions: %v", err),
			}
		}
	}

	// Prepare trade data for response
	var tradeData []map[string]interface{}
	for _, trade := range trades {
		tradeData = append(tradeData, map[string]interface{}{
			"trade_id": trade.TradeID,
			"price":    trade.Price,
			"quantity": trade.Quantity,
		})
	}

	return SubmitOrderResult{
		Success: true,
		OrderID: order.OrderID,
		Status:  string(order.Status),
		Trades:  tradeData,
	}
}

// CancelOrder cancels an existing order
func (om *OrderManager) CancelOrder(orderID string, userID uint) map[string]interface{} {
	var order core.Order
	if err := om.db.Where("order_id = ? AND user_id = ? AND status = ?", 
		orderID, userID, core.OrderStatusPending).First(&order).Error; err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   "Order not found or cannot be cancelled",
		}
	}

	order.Status = core.OrderStatusCancelled
	if err := om.db.Save(&order).Error; err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   fmt.Sprintf("Failed to cancel order: %v", err),
		}
	}

	return map[string]interface{}{
		"success": true,
		"message": "Order cancelled successfully",
	}
}

// GetUserOrders returns user's orders
func (om *OrderManager) GetUserOrders(userID uint, limit int) ([]map[string]interface{}, error) {
	var orders []core.Order
	query := om.db.Where("user_id = ?", userID).Order("created_at desc")
	if limit > 0 {
		query = query.Limit(limit)
	}
	
	if err := query.Find(&orders).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch orders: %w", err)
	}

	var result []map[string]interface{}
	for _, order := range orders {
		orderData := map[string]interface{}{
			"order_id":        order.OrderID,
			"contract_id":     order.ContractID,
			"side":            string(order.Side),
			"order_type":      string(order.OrderType),
			"quantity":        order.Quantity,
			"price":           order.Price,
			"status":          string(order.Status),
			"filled_quantity": order.FilledQuantity,
			"avg_fill_price":  order.AvgFillPrice,
			"created_at":      order.CreatedAt.Format(time.RFC3339),
			"updated_at":      order.UpdatedAt.Format(time.RFC3339),
		}
		result = append(result, orderData)
	}

	return result, nil
}

// GetUserTrades returns user's trade history
func (om *OrderManager) GetUserTrades(userID uint, limit int) ([]map[string]interface{}, error) {
	var trades []core.Trade
	
	// Join with orders to filter by user
	query := om.db.Joins("JOIN orders ON (trades.buy_order_id = orders.order_id OR trades.sell_order_id = orders.order_id)").
		Where("orders.user_id = ?", userID).
		Order("trades.trade_time desc")
	
	if limit > 0 {
		query = query.Limit(limit)
	}
	
	if err := query.Find(&trades).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch trades: %w", err)
	}

	var result []map[string]interface{}
	for _, trade := range trades {
		tradeData := map[string]interface{}{
			"trade_id":    trade.TradeID,
			"contract_id": trade.ContractID,
			"quantity":    trade.Quantity,
			"price":       trade.Price,
			"trade_time":  trade.TradeTime.Format(time.RFC3339),
		}
		result = append(result, tradeData)
	}

	return result, nil
}

// GetUserPositions returns user's current positions
func (om *OrderManager) GetUserPositions(userID uint) ([]map[string]interface{}, error) {
	var positions []core.Position
	if err := om.db.Where("user_id = ?", userID).Find(&positions).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch positions: %w", err)
	}

	var result []map[string]interface{}
	for _, pos := range positions {
		posData := map[string]interface{}{
			"position_id":        pos.ID,
			"contract_id":        pos.ContractID,
			"quantity":           pos.Quantity,
			"avg_entry_price":    pos.AvgEntryPrice,
			"unrealized_pnl":     pos.UnrealizedPnL,
			"realized_pnl":       pos.RealizedPnL,
			"margin_requirement": pos.MarginRequirement,
			"last_updated":       pos.LastUpdated.Format(time.RFC3339),
		}
		result = append(result, posData)
	}

	return result, nil
}

// updatePositions updates user positions based on executed trades
func (om *OrderManager) updatePositions(userID uint, trades []core.Trade) error {
	for _, trade := range trades {
		// Find existing position
		var position core.Position
		err := om.db.Where("user_id = ? AND contract_id = ?", userID, trade.ContractID).First(&position).Error
		
		// Determine if this is a buy or sell
		var order core.Order
		if trade.BuyOrderID != nil {
			if err := om.db.Where("order_id = ?", *trade.BuyOrderID).First(&order).Error; err != nil {
				continue
			}
		} else if trade.SellOrderID != nil {
			if err := om.db.Where("order_id = ?", *trade.SellOrderID).First(&order).Error; err != nil {
				continue
			}
		} else {
			continue
		}

		tradeQuantity := trade.Quantity
		if order.Side == core.OrderSideSell {
			tradeQuantity = -trade.Quantity
		}

		if err == nil {
			// Update existing position
			oldQuantity := position.Quantity
			oldValue := oldQuantity * position.AvgEntryPrice
			newValue := tradeQuantity * trade.Price

			position.Quantity += tradeQuantity

			if position.Quantity != 0 {
				position.AvgEntryPrice = (oldValue + newValue) / position.Quantity
			} else {
				// Position closed
				position.RealizedPnL += oldValue + newValue
				position.AvgEntryPrice = 0
			}

			position.LastUpdated = time.Now()
			if err := om.db.Save(&position).Error; err != nil {
				return fmt.Errorf("failed to update position: %w", err)
			}
		} else {
			// Create new position
			var contract core.Contract
			if err := om.db.First(&contract, trade.ContractID).Error; err != nil {
				continue
			}

			position = core.Position{
				UserID:            userID,
				ContractID:        trade.ContractID,
				Quantity:          tradeQuantity,
				AvgEntryPrice:     trade.Price,
				MarginRequirement: contract.InitialMargin,
				LastUpdated:       time.Now(),
			}

			if err := om.db.Create(&position).Error; err != nil {
				return fmt.Errorf("failed to create position: %w", err)
			}
		}
	}

	return nil
}

// UpdatePositionPnL updates unrealized P&L for all positions based on current market prices
func (om *OrderManager) UpdatePositionPnL(currentPrices map[uint]float64) error {
	var positions []core.Position
	if err := om.db.Find(&positions).Error; err != nil {
		return fmt.Errorf("failed to fetch positions: %w", err)
	}

	for _, position := range positions {
		if currentPrice, exists := currentPrices[position.ContractID]; exists {
			if position.Quantity != 0 {
				position.UnrealizedPnL = (currentPrice - position.AvgEntryPrice) * position.Quantity
				position.LastUpdated = time.Now()
				
				if err := om.db.Save(&position).Error; err != nil {
					return fmt.Errorf("failed to update position P&L: %w", err)
				}
			}
		}
	}

	return nil
}